Mtracy/sub navigation

mrtracy opened the pull request:
Add per-page "tab" navigation from admin UI:

![screen shot 2015-08-06 at 3 39 30 pm](https://cloud.githubusercontent.com/assets/6969858/9121619/7178e44e-3c51-11e5-9edc-f0c115749cbb.png)
![screen shot 2015-08-06 at 3 39 39 pm](https://cloud.githubusercontent.com/assets/6969858/9121621/740ab82c-3c51-11e5-8235-4ea40398bc8b.png)

Commit:
Admin UI: Add Page Tab Navigation Bar

In the latest design wireframes, certain UI pages are composed of multiple
"Tabs" - for example, the page for a single Node will have an "Overview" tab and
a "Graphs" tab.

Navigation between these tabs is accomplished via a Navigation Bar component,
which is the same component used for the main navigation bar.

Finally, the content on the various page tabs has been restructured to more
closely resemble the current wireframes.

BramGruneir commented on ui/ts/components/table.ts:
```diff
@@ -133,7 +133,7 @@ module Components {
             {
               onclick: (e: any): void => this.SetSortColumn(col),
-              className: this.IsSortColumn(col) ? sortClass : null
+              className: this.IsSortColumn(col) ? sortClass : ""
```
why change this?

mrtracy replied:
it was actually assigning `class="null"` to the html elements. With this, the class is actually blank.

BramGruneir replied:
cool

BramGruneir commented on ui/ts/pages/nodes.ts:
```diff
@@ -37,14 +40,14 @@ module AdminViews {
           {
             title: "Address",
-            view: (status: NodeStatus): string => status.desc.address.address,
+            view: (status: NodeStatus): string => status.desc.address.string,
```
should we address it with ["string"] ?

mrtracy replied:
I should just take the quotes off of the proto, its not a reserved word.

mrtracy replied:
Scratch that: I just changed the proto itself on the backend so the field is "address" again.

BramGruneir commented on ui/ts/pages/nodes.ts:
```diff
@@ -120,15 +123,15 @@ module AdminViews {
-          m("h2", "Nodes List"),
-          m(".section", ctrl.PrimaryStats()),
-          m(".stats-table", Components.Table.create(comparisonData))
+        return m(".page", [
```
does this assume a div?

mrtracy replied:
yes.

BramGruneir replied:
cool

BramGruneir commented on ui/ts/pages/stores.ts:
```diff
@@ -138,7 +141,24 @@ module AdminViews {
+          }
+        ];
+
+        private static isActive: (targ: NavigationBar.Target) => boolean = (t: NavigationBar.Target) => {
```
shouldn't this isactive be part of the navigationBar targets instead of the navigation bar itself? Each one should have its own isActive based on the route?

BramGruneir replied:
it seems unnecessary to have to redefine isActive for each use of the navBar.  Is there no single function to handle it?

mrtracy replied:
Well it made sense to me, my thought was that a single nav bar was going to switch between discrete options that are at roughly the same "level" of the page - the top navigation bar determines the base route, while the subpage navigation bar just determines a single parameter of the route.

It made sense to me.

mrtracy replied:
And no, there's definitely no single function to handle it - there are two nav bars on the screen and thus two active links.

BramGruneir replied:
Instead of one function that returns a target, how about doing the opposite, have a function on each target, that determines if it is the active one?
This way, you could make it a readOnlyProp that depends on the route.

BramGruneir replied:
I think it would be cleaner.  But up to you.

mrtracy replied:
The route is not one of our properties, it's a mithril property - we can't depend on it with our computed properties.

Also, 'isActive' doesn't return the active target - it accepts a target, and returns true if it is active.  It's only used to add a css class to the active link.

BramGruneir replied:
As discussed offline, keep it this way.

tamird commented on ui/ts/models/proto.ts:
```diff
@@ -92,7 +92,7 @@ module Models {
     export interface Addr {
       network: string;
-      address: string;
+      "string": string;
```
as discussed in person: probably better to rename this in the proto

mrtracy replied:
Done, proto has been renamed.

Commit:
UnresolvedAddress.string_field -> address_field

The field name of this structure was changed by #1976, but it has since been
decided that "address" is a more accurate name than "string".

BramGruneir commented:
LGTM

tamird commented:
Disregard comma comments, see https://github.com/palantir/tslint/issues/557
